utilities/process/diffusion_utils.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+
+
+def angle_between(v1, v2):
+    """ Returns the angle in radians between vectors 'v1' and 'v2'::
+
+            >>> angle_between((1, 0, 0), (0, 1, 0))
+            1.5707963267948966
+            >>> angle_between((1, 0, 0), (1, 0, 0))
+            0.0
+            >>> angle_between((1, 0, 0), (-1, 0, 0))
+            3.141592653589793
+    """
+    nv1 = np.linalg.norm(v1)
+    nv2 = np.linalg.norm(v2)
+    if nv1 == 0 or nv2 == 0:
+        return 0
+    v1_u = v1 / nv1
+    v2_u = v2 / nv2
+    return np.degrees(np.arccos(np.clip(np.dot(v1_u, v2_u), -1.0, 1.0)))
+
+def find_shells(bvals_raw, rtol=0.01, atol=1):
+    """ Automatically finds shells from a bvalue file.
+    
+    Given a raw list of bvalues and optional relative and absolute tolerances,
+    return the unique shells, shell indices, and b0 image indices.
+
+    Parameters
+    ------------
+    bvals_raw : list
+        List of bvalues as acquired
+    rtol : float
+        Relative tolerance of bvalues
+    atol : float
+        Absolute tolerance of bvalues
+
+    Returns
+    ----------
+    shells : list
+        The mean values of the shells
+    indices : list
+        The indices of the shells in the original bvalue list
+    b0 : list
+        A boolean array of locations of b0 values
+    """
+    # if we know the number of values in each shell we could use np.partition
+    # but I don't think that's very useful.
+
+    bvals_raw = np.asarray(bvals_raw)
+    assert len(bvals_raw.shape) == 1, "Must be a 1D array"
+
+    # group the bvalues based on tolerances
+    bvals_indices = np.argsort(bvals_raw)  # sort order
+    bvals = bvals_raw[bvals_indices]
+    index = np.zeros_like(bvals, dtype=int)
+    idx_lower = 0
+    for idx_higher in range(1, len(bvals)):
+        if np.allclose(bvals[idx_lower], bvals[idx_higher], rtol, atol):
+            index[idx_higher] = index[idx_lower]
+        else:
+            index[idx_higher] = index[idx_lower] + 1
+            idx_lower = idx_higher
+
+    # invert the sort to return the shell indices of the unsorted bvalues
+    bvals_inverse_sort = np.empty_like(bvals_indices)
+    bvals_inverse_sort[bvals_indices] = np.arange(bvals_indices.size)
+
+    # find the mean b-values of the shells
+    shells = [np.mean(bvals[index==idx]) for idx in set(index)]
+    indices = index[bvals_inverse_sort]
+    return shells, indices, indices<atol #, bvals_indices
+
+def find_directions(bvecs_raw, zero_indices, atol=1, rotationally_symmetric=True):
```

Comment:
Looks good. Are we doing a lot with vectors?

Reply:
Not really right now. I was thinking some algorithms will want to operate on a single direction at a time so that's why I wrote this, to be able to automatically split those up.